### PR TITLE
Everlong testing harness

### PR DESCRIPTION
**Pull request by mcclurejt**
# Resolved Issues

- closes https://github.com/delvtech/hyperdrive/issues/1083

# Description

Includes a base test harness for Everlong that extends the existing `HyperdriveTest.sol` harness.

# Review Checklists

## @jalextowle 

- [ ] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [ ] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

## @jrhea  

- [ ] **Tokens**
    - [ ] Do all `approve` calls use `forceApprove`?
    - [ ] Do all `transfer` calls use `safeTransfer`?
    - [ ] Do all `transferFrom` calls use `msg.sender` as the `from` address?
        - [ ] If not, is the function access restricted to prevent unauthorized
              token spend?
- [ ] **Low-level calls (`call`, `delegatecall`, `staticcall`, `transfer`, `send`)**
    - [ ] Is the returned `success` boolean checked to handle failed calls?
    - [ ] If using `delegatecall`, are there strict access controls on the
          addresses that can be called? It shouldn't be possible to `delegatecall`
          arbitrary addresses, so the list of possible targets should either be
          immutable or tightly controlled by an admin.
- [ ] **Reentrancy**
    - [ ] Are functions that make external calls or transfer ether marked as `nonReentrant`?
        - [ ] If not, is there documentation that explains why reentrancy is
              not a concern or how it's mitigated?
- [ ] **Gas Optimizations**
    - [ ] Is the logic as simple as possible?
    - [ ] Are the storage values that are used repeatedly cached in stack or
          memory variables?
    - [ ] If loops are used, are there guards in place to avoid out-of-gas
          issues?
- [ ] **Visibility**
    - [ ] Are all `payable` functions restricted to avoid stuck ether?
- [ ] **Math**
    - [ ] Is all of the arithmetic checked or guarded by if-statements that will
          catch underflows?
    - [ ] If `Safe` functions are altered, are potential underflows and
          overflows caught so that a failure flag can be thrown?
    - [ ] Are all of the rounding directions clearly documented?
- [ ] **Testing**
    - [ ] Are there new or updated unit or integration tests?
    - [ ] Do the tests cover the happy paths?
    - [ ] Do the tests cover the unhappy paths?
    - [ ] Are there an adequate number of fuzz tests to ensure that we are
          covering the full input space?

### CODE REVIEW

File: contracts/src/everlong/EverlongAdmin.sol [hunk 1]
```diff
@@ -1,12 +1,16 @@
 // SPDX-License-Identifier: Apache-2.0
 pragma solidity 0.8.20;
 
-import { IEverlongAdmin } from "../interfaces/IEverlongAdmin.sol";
+import { IEverlongAdmin } from "contracts/src/interfaces/IEverlongAdmin.sol";
```
Comment (jalextowle) [inline contracts/src/everlong/EverlongAdmin.sol:4]: Why are we making these absolute? Everywhere else in the project, we use relative for contracts in `contracts/src` and absolute elsewhere. The reason why we do this is annoying and historical, but I don't want to change unless we change all of the files.

File: test/utils/EverlongTest.sol [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import { HyperdriveTest } from "test/utils/HyperdriveTest.sol";
+import { IEverlong } from "contracts/src/interfaces/IEverlong.sol";
+import { Everlong } from "contracts/src/everlong/Everlong.sol";
+
+contract EverlongTest is HyperdriveTest {
+    IEverlong internal everlong;
+
+    function setUp() public virtual override {
+        super.setUp();
+        vm.startPrank(alice);
```
Comment (jalextowle) [inline test/utils/EverlongTest.sol:15]: What's the point of this? `deployEverlong` immediately switches to the `deployer`.

File: test/utils/EverlongTest.sol [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+pragma solidity 0.8.20;
+
+import { VmSafe } from "forge-std/Vm.sol";
+import { BaseTest } from "test/utils/BaseTest.sol";
+import { HyperdriveTest } from "test/utils/HyperdriveTest.sol";
+import { IEverlong } from "contracts/src/interfaces/IEverlong.sol";
+import { Everlong } from "contracts/src/everlong/Everlong.sol";
+
+contract EverlongTest is HyperdriveTest {
+    IEverlong internal everlong;
```
Comment (jalextowle) [inline test/utils/EverlongTest.sol:11]: I love this. Using interfaces from the start will make our lives much easier.

**Review by jalextowle — COMMENTED** [review]
It's looking good. Just some style nits.

File: contracts/src/everlong/EverlongAdmin.sol (hunk 1 above)
Comment (mcclurejt) [reply contracts/src/everlong/EverlongAdmin.sol:4]: Got it, didn't make the connection on where rel/abs paths were being used, will put it back to relative

File: contracts/src/everlong/EverlongERC4626.sol [hunk 4]
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: Apache-2.0
 pragma solidity 0.8.20;
 
-// import { IEverlongERC4626 } from "../interfaces/IEverlongERC4626.sol";
+import { IHyperdrive } from "contracts/src/interfaces/IHyperdrive.sol";
```
[the remark this answers is not in the document]
Comment (mcclurejt) [reply contracts/src/everlong/EverlongERC4626.sol]: https://github.com/delvtech/hyperdrive/pull/1097/files#diff-90f83f6ad18083e73ae495e1be72e15412314e5db271c2ec8121904305a82fbdL54

**Pull request closed without merging**